container-search/src/main/java/com/yahoo/search/dispatch/LoadBalancer.java

```diff
@@ -3,25 +3,108 @@
 import com.yahoo.search.Query;
 import com.yahoo.search.dispatch.SearchCluster.Group;
 
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
 import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 public class LoadBalancer {
+    // The implementation here is a simplistic least queries in flight + round-robin load balancer
+    // TODO: consider the options in com.yahoo.vespa.model.content.TuningDispatch
 
-    private final SearchCluster searchCluster;
+    private final static Logger log = Logger.getLogger(LoadBalancer.class.getName());
+
+    private final boolean isInternallyDispatchable;
+    private final List<GroupSchedule> scoreboard;
+    private int needle = 0;
 
     public LoadBalancer(SearchCluster searchCluster) {
-        this.searchCluster = searchCluster;
+        if (searchCluster == null) {
+            this.isInternallyDispatchable = false;
+            this.scoreboard = null;
+            return;
+        }
+        this.isInternallyDispatchable = (searchCluster.groupSize() == 1);
+        this.scoreboard = new ArrayList<>(searchCluster.groups().size());
+
+        for (Group group : searchCluster.groups().values()) {
+            scoreboard.add(new GroupSchedule(group));
+        }
+        Collections.shuffle(scoreboard);
     }
 
     public Optional<Group> getGroupForQuery(Query query) {
```

Comment:
The Javadoc for this should say that callers *must* call releaseGroup exactly once per invocation of this which does not return empty. Imho it would also help to call it takeGroup as take/release is symmetric (in my mind anyway).